dtfs-central-api/src/v1/controllers/utilisation-report-service/put-utilisation-report-status.controller/index.ts

```diff
@@ -0,0 +1,97 @@
+
+export type PutUtilisationReportStatusRequest = CustomExpressRequest<{
+  reqBody: {
+    user?: {
+      _id?: TfmSessionUser['_id'];
+    };
+    reportsWithStatus?: ReportWithStatus[];
+  };
+}>;
+
+const assertReportWithStatusIsPopulated = (reportWithStatus: ReportWithStatus) => {
+  const { status, reportId } = reportWithStatus;
+  if (!status || !reportId) {
+    throw new InvalidPayloadError("Request body item 'reportsWithStatus' supplied does not match required format");
+  }
+};
+
+const executeEventHandlerWithQueryRunner = async (
+  reportWithStatus: ReportWithStatus,
+  transactionEntityManager: EntityManager,
+  requestSource: DbRequestSource,
+): Promise<void> => {
+  const stateMachine = await UtilisationReportStateMachine.forReportId(reportWithStatus.reportId);
+
+  if (reportWithStatus.status === 'RECONCILIATION_COMPLETED') {
+    await stateMachine.handleEvent({
+      type: 'MANUALLY_SET_COMPLETED',
+      payload: {
+        requestSource,
+        transactionEntityManager,
+      },
+    });
+    return;
+  }
+  await stateMachine.handleEvent({
+    type: 'MANUALLY_SET_INCOMPLETE',
+    payload: {
+      requestSource,
+      transactionEntityManager,
+    },
+  });
+};
+
+export const putUtilisationReportStatus = async (req: PutUtilisationReportStatusRequest, res: Response) => {
+  try {
+    const { reportsWithStatus, user } = req.body;
+
+    if (!user || !user._id) {
+      throw new InvalidPayloadError("Request body item 'user' supplied does not match required format");
+    }
+
+    const requestSource: DbRequestSource = {
+      platform: 'TFM',
+      userId: user._id.toString(),
+    };
+
+    if (!reportsWithStatus) {
+      throw new InvalidPayloadError("Request body item 'reportsWithStatus' supplied does not match required format");
+    }
+    reportsWithStatus.forEach(assertReportWithStatusIsPopulated);
+
+    const queryRunner = SqlDbDataSource.createQueryRunner();
+    await queryRunner.connect();
+
+    await queryRunner.startTransaction();
+    try {
+      const transactionEntityManager = queryRunner.manager;
+      await Promise.all(
+        reportsWithStatus.map((reportWithStatus) => executeEventHandlerWithQueryRunner(reportWithStatus, transactionEntityManager, requestSource)),
+      );
+      await queryRunner.commitTransaction();
+    } catch (error) {
+      await queryRunner.rollbackTransaction();
+      throw new TransactionFailedError();
+    } finally {
+      await queryRunner.release();
+    }
+
+    return res.sendStatus(HttpStatusCode.Ok);
+  } catch (error) {
+    console.error('Error updating utilisation report statuses:', error);
+    if (error instanceof ApiError) {
+      return res.status(error.status).send(`Failed to update utilisation report statuses: ${error.message}`);
+    }
+    return res.status(HttpStatusCode.InternalServerError).send(`Failed to update utilisation report statuses`);
+  }
```

Comment:
> I don't think there is a way for us to get around this, so it is unfortunately something we just need to accept whenever we perform transactions

I think we will probably want to have quite a lot of stuff using transactions and we need a solution that works with our repo objects, even if that means adapting our repo objects. As I say I'm not an expert on this by any means but I do think it should be possible to come up with a pattern that works nicely with transactions or I'm sure typeORM would not be as popular as it is.

This [article](https://medium.com/@dev.muhammet.ozen/advanced-transaction-management-with-nestjs-typeorm-43a839363491) is sadly using nest.js so the code has to be viewed in that context, it obviously won't look identical for us at all, but looking at the start of the advanced section under the repositories header, they actually pass a dataSource to their repo constructor. Perhaps we can take some inspiration from here to adapt our repo implementation?

Reply:
The way we define our repository is just to extend the repository which exists on our `SqlDbDataSource` connection, so you can do the same thing with the transaction entity manager:

```typescript
const UtilisationReportRepo = transactionEntityManager.manager.getRepository(UtilisationReportEntity).extend({ ... });
```

However, as we don't need it right now, I'd argue YAGNI may apply as we may not need our more complex repo methods for any transactions that we perform

Reply:
Originally when I looked at this I mostly thought 'this is a one off transaction thats going to go in the bin soon' hence why I was quite relaxed to just use a queryRunner and not worry too much. I still can't tell in TFM if we are going to have a tonne or almost no transactions aha - although this final part you're describing Josh, is this just literally creating a new repo extension on the transactions version of the UtilisationReportEntity that adds all our existing ones on?
Have you had a chance to test if this works etc.?

Reply:
Yes - this is just creating a new repo extension, but using all our methods that we have already defined on the original `UtilisationReportRepo`. I haven't tested it at runtime, but typescript sees all the methods defined on our `UtilisationReportRepo` in `TransactionRepo` (as defined above), so I don't see why it wouldn't work at runtime

Reply:
Interesting! For this place it would be nice to:
- Pull all the transaction related stuff into a separate function, where that lives might be a bit hard, I'd expect it to be in something like a utilisation-report-service but we don't really have services.
- That function would start the transaction and then extend to have the extra repo classes
- Calls whatever repo methods it needs

I'm assuming that we can't have a constant transaction manager that is extended in methods because each 'transactionManager' will be a different one that needs to be extended individually.

We could have  amethod as you said in the UtilisationReportRepo that takes in a transactionEntityManager, extends it, and returns the new transactionManager which might be nice!

Thoughts from anyone?

Reply:
Had a small discussion, for now can we just extract the transaction style logic to a separate function and we can revisit if we end up having a lot of transactions coming in in TFM release 6

Reply:
Just as a closing comment, I've also added some docstrings to all the functions in this file